File: javascript/hard/bfsNodeDistances.js
// hard, BFS
// currently only passes first test of 8

class Node {
  constructor(value) {
    this.value = value;
    this.edges = [];
    this.distanceFromS = -1;
  }
  addChild(node) {
    this.edges.push(node);
  }
  setDistance(x) {
    this.distanceFromS = x;
  }
}

function bfs(n, m, edges, s) {
  // create array of all input nodes
  const nodes = [];
  for (let i = 0; i < n; i++) {
    nodes.push(new Node(i + 1));
  }

  // create edges for all nodes
  for (let i = 0; i < m; i++) {
    nodes[edges[i][0] - 1].addChild(nodes[edges[i][1] - 1]);
    nodes[edges[i][1] - 1].addChild(nodes[edges[i][0] - 1]);
  }

  let nodesVisited = [s];
  function checkEdgesAndSetDistance(node) {
    for (let i = 0; i < node.edges.length; i++) {
      nodesVisited.push(node.edges[i].value);

      const distance = node.edges[i].distanceFromS;
      if (distance === -1) {
        node.edges[i].setDistance(6);
      } else {
        node.edges[i].setDistance(distance + 6);
      }

      if (nodesVisited.find((j) => j === node.value)) continue;
      else checkEdgesAndSetDistance(node.edges[i]);
    }
  }
  nodes[s - 1].setDistance = 0;
  checkEdgesAndSetDistance(nodes[s - 1]);

  const distances = nodes.map((node) => node.distanceFromS);
  distances.splice(s - 1, 1);
  return distances;
}

File: python/hard/bfs_node_distances.py
# hard, BFS/graph
# https://www.hackerrank.com/challenges/bfsshortreach/problem

# solution on 01-26-23
from collections import deque

def bfs(n, m, edges, s):
    graph = {}
    for i in range(n):
        graph[i+1] = set()
    for n1, n2 in edges:
        graph[n1].add(n2)
        graph[n2].add(n1)
    
    distance_queue = deque([(s, 0)])
    nodes_visited = {}
    distances = [-1] * n
    
    while distance_queue:
        node, dist = distance_queue.popleft()
        if node not in nodes_visited:
            nodes_visited[node] = 1
            distances[node-1] = dist
            for neighbor_node in graph[node]:
                distance_queue.append([neighbor_node, dist+6])
    
    return distances[:s-1] + distances[s:]


# solution on 07-10-23 (did not realize i had already done this problem)
def bfs(n, m, edges, s):
    # create graph (each index contains a nodes neighbors)
    graph = [[] for _ in range(n)]
    for a, b in edges:
        graph[a-1].append(b)
        graph[b-1].append(a)
    # record distances
    distances = [-1] * n
    distances[s-1] = 0
    # use queue to traverse graph with BFS
    queue = [s]
    while queue:
        node = queue.pop(0)
        for neighbor in graph[node-1]:
            if distances[neighbor-1] == -1:
                distances[neighbor-1] = distances[node-1] + 6
                queue.append(neighbor)
    return distances[:s-1] + distances[s:]
